misc/metrics: Add `protocols` label to address-specific metrics

John-LittleBearLabs opened the pull request:
## Description

Previously, we would only track the metrics like the number of open connections. With this patch, we extend these metrics with a `protocols` label that contains a "protocol stack". A protocol stack is a multi-address with all variable parts removed. For example, `/ip4/127.0.0.1/tcp/1234` turns into `/ip4/tcp`.

Resolves https://github.com/libp2p/rust-libp2p/issues/2758.

## Links to any relevant issues

## Open Questions

~The use of Family appends `_total` to the metric name. Is this desirable?~

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

John-LittleBearLabs committed:
Getting started

John-LittleBearLabs committed:
Merge remote-tracking branch 'jt/master' into addrstak_metric

John-LittleBearLabs committed:
Merge remote-tracking branch 'origin/master' into addrstak_metric

John-LittleBearLabs committed:
Post merge fixup and re-check.

thomaseizinger commented on misc/metrics/Cargo.toml:
```diff
@@ -21,11 +21,11 @@ dcutr = ["libp2p-dcutr"]
 libp2p-core = { version = "0.37.0", path = "../../core" }
 libp2p-dcutr =  { version = "0.7.0", path = "../../protocols/dcutr", optional = true }
-libp2p-identify = { version = "0.40.0", path = "../../protocols/identify", optional = true }
+libp2p-identify = { path = "../../protocols/identify", optional = true }
```
Please retain the versions here.

thomaseizinger commented on Cargo.toml:
```diff
@@ -109,11 +109,12 @@ libp2p-swarm-derive = { version = "0.30.1", path = "swarm-derive" }
 libp2p-wasm-ext = { version = "0.37.0", path = "transports/wasm-ext", optional = true }
 libp2p-yamux = { version = "0.41.0", path = "muxers/yamux", optional = true }
-multiaddr = { version = "0.14.0" }
+multiaddr = { path = "../rust-multiaddr" }
```
Can we for now have the `protocol_stack` function in here? Upstreaming it right away doesn't feel right as we may have to iterate on it for a bit.

thomaseizinger reviewed (commented):
Thanks for working on this!

John-LittleBearLabs committed:
Pulling protocol_stack functionality into this repo, by request.

thomaseizinger commented on misc/metrics/Cargo.toml:
```diff
@@ -23,8 +23,8 @@ libp2p-core = { version = "0.37.0", path = "../../core" }
-libp2p-ping = { version = "0.40.0", path = "../../protocols/ping", optional = true }
-libp2p-relay =  { version = "0.13.0", path = "../../protocols/relay", optional = true }
+libp2p-ping = { version = "0.40.0",  path = "../../protocols/ping", optional = true }
+libp2p-relay =  {  version = "0.13.0", path = "../../protocols/relay", optional = true }
```
Please undo this formatting :)

thomaseizinger commented on misc/metrics/examples/metrics/main.rs:
```diff
@@ -109,8 +113,23 @@ fn main() -> Result<(), Box<dyn Error>> {
 /// and can be observed via the metrics.
-#[derive(NetworkBehaviour, Default)]
+#[derive(NetworkBehaviour)]
+// #[behaviour(out_event = "Event")]
```
```suggestion
```

thomaseizinger commented on misc/metrics/src/identify.rs:
```diff
@@ -27,6 +27,8 @@ use prometheus_client::registry::Registry;
 use std::iter;
 use std::sync::{Arc, Mutex};
+use prometheus_client::metrics::family::Family;
+use super::protocol_stack::protocol_stack;
```
Please use `crate::` for importing crate-local functions.

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,13 @@
+    format!("{}", proto).split("/").skip(1).next().map(str::to_owned)
+}
+
+pub(crate) fn protocol_stack(ma: &Multiaddr) -> Vec<String> {
```
```suggestion
pub fn protocol_stack(ma: &Multiaddr) -> Vec<String> {
```

The entire module is crate-private so we don't need this! :)

thomaseizinger replied:
Can we have a few tests for this please?

John-LittleBearLabs replied:
Honestly I was trying to go minimal here because in my mind the "real" implementation is still [in multiaddr](https://github.com/multiformats/rust-multiaddr/pull/60). But yeah I could put a test here too.

thomaseizinger replied:
Oh I wasn't aware of that PR, sorry about that!

We'd still probably want a test that shows that we construct the correct stack as it is described in the issue. It makes review easier because I can see the output of the function without having to run or brain-compile it :)

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,13 @@
+
+fn tag(proto: Protocol) -> Option<String> {
+    format!("{}", proto).split("/").skip(1).next().map(str::to_owned)
+}
```
Can we have helpers at the bottom of the file please?

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,13 @@
+//  This does imply changing the keys from Vec<String> to Vec<'static str>
+
+fn tag(proto: Protocol) -> Option<String> {
+    format!("{}", proto).split("/").skip(1).next().map(str::to_owned)
```
It would be great to avoid these allocations here. It creates some duplication but we could `match` on all variants and return a `&'static str` instead of using `Display`.

Metrics should be cheap to record so I think this is worth it.

John-LittleBearLabs replied:
Yeah the rust-multiaddr implementation is an `Iterator<Item = &'static str>` because I had the same basic concern. I can copy-paste more.

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,13 @@
+
+//TODO: remove this file and replace with calls to Multiaddr::protocol_stack
+//  once that lands : https://github.com/multiformats/rust-multiaddr/pull/60
+//  This does imply changing the keys from Vec<String> to Vec<'static str>
```
To signal that this is function has upstream potential, we could model it as an extension trait:

```rust
trait MultiaddrExt {
	fn protocol_stack(&self) -> String;
}
```

John-LittleBearLabs replied:
I'm not sure how that would look. We'd be writing an `impl MultiaddrExt for Multiaddr` in both repos? Or a helper struct in this repo that wraps around Multiaddr?

thomaseizinger replied:
Not quite. The idea I wanted to convey is that from a user PoV, an extension trait makes the interface look like it is already on `Multiaddr`, i.e. we are _extending_ `Multiaddr` with functionality that we would like to have there. Once you do `impl MultiaddrExt for Multiaddr`, you can just call `multiaddr.protocol_stack()` as if `Multiaddr` would natively provide that function.

That way, call-sites don't have to change once we do actually upstream it. At that point, the trait can be deleted and the function just goes onto a regular `impl Multiaddr` block.

Semantically, it doesn't make a difference but I find that an extension trait conveys a different intent compared to a module-private utility function :)

I don't feel strongly about it though so feel free to go with whatever!

thomaseizinger replied:
Don't mind the above, I didn't realize you were already creating the function upstream. All of this was so that we don't _need_ to create it upstream but can instead iterate on it here first.

thomaseizinger commented on misc/metrics/src/swarm.rs:
```diff
@@ -156,8 +157,9 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
-                self.connections_incoming.inc();
+            libp2p_swarm::SwarmEvent::IncomingConnection { send_back_addr, .. } => {
+                let labels = protocol_stack(send_back_addr);
+                self.connections_incoming.get_or_create(&labels).inc();
```
I don't think we want to create one label per protocol. Instead we want one label per entire stack, i.e. the labels here should be all concatenated.

This is probably best to happen inside the `protocol_stack` function so it only returns a `String`.

John-LittleBearLabs replied:
That actually raises a risk of collision (not right now, but a potential future tag) unless a delimiter is used. The most obvious delimiter would be "/".

After we go to `&str` it should be a similar number of allocations between `Vec<&'static str>` and `String`, correct? Are you thinking comparisons and hashing is faster with String or something of that nature?

thomaseizinger replied:
> That actually raises a risk of collision (not right now, but a potential future tag) unless a delimiter is used. The most obvious delimiter would be "/".

Yeah, the original issue uses `/` as a delimiter: https://github.com/libp2p/rust-libp2p/issues/2758

> After we go to `&str` it should be a similar number of allocations between `Vec<&'static str>` and `String`, correct? Are you thinking comparisons and hashing is faster with String or something of that nature?

The allocations I was concerned with were the ones were we allocate a new string _for each_ protocol in the multiaddr. Constructing a new, single string will require one allocation. I don't think we can get rid of that one.

This should do the trick[^1] (note the leading `/` at `/tcp`:

```rust
let stack = addr
	.iter()
	.map(|p| match p {
		Protocol::Tcp(_) => "/tcp",
	 	// etc
	})
	.collect::<String>();
```

[^1]: Looking at the `collect` impl, we will still allocate multiple times depending on the length of the multiaddr though: https://doc.rust-lang.org/src/alloc/string.rs.html#2073

thomaseizinger commented on misc/metrics/src/swarm.rs:
```diff
@@ -246,6 +248,7 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
 #[derive(Encode, Hash, Clone, Eq, PartialEq)]
 struct ConnectionEstablishedLabels {
     role: Role,
+
```
Please remove this unneeded reformatting.

thomaseizinger reviewed (changes requested):
Thanks! A few comments.

John-LittleBearLabs committed:
A couple of suggestions from PR

John-LittleBearLabs committed:
Merge remote-tracking branch 'origin/master' into addrstak_metric

John-LittleBearLabs committed:
Label wrapper

John-LittleBearLabs committed:
MultiaddrExt

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+            result.push('/');
+            result.push_str(proto_tag);
+        }
+        result
```
Nice work!

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+     fn from(ma: &Multiaddr) -> Self {
+        Self { address_stack: ma.protocol_stack() }
+    }
+}
```
I don't find these to be particularly intuitive. Just because something can be expressed with `From` doesn't mean it should use `From`.

I'd be in favor of just saying:

```rust
impl Label {
	pub fn for_multi_address(addr: &Multiaddr) -> Self {}
}
```

This makes it a lot more clear on the call-site, what is happening. `into()` is very in-transparent. One of the best uses of `From` and `Into` (as with all traits actually) is to create APIs that are generic over these capabilities, like:

```rust
fn do_something(value: impl Into<Label>) {
	let label = value.into();

	// Do something with `Label`
}
```

If you have to call `into()` yourself - as we do in `identify.rs` - the value of using a standardized trait to do your conversion is practically zero. I'd actually claim it is harmful because it obfuscates what is happening which is why I'd prefer a more concrete function that conveys more intent.

Hope that makes sense :)

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+        let mut buf = Vec::new();
+        actual.encode(&mut buf).expect("encode failed");
+        let actual = String::from_utf8(buf).expect("invalid utf-8?");
+        assert_eq!(actual, "address_stack=\"/ip6/tcp/wss/p2p\"");
```
You can use raw strings (`r#""#`) to avoid the escaping.

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+#[derive(Encode, Hash, Clone, Eq, PartialEq)]
+pub struct Label {
+    address_stack: String
+}
```
Given that this is our new public API, can it go at the top of the file please?

thomaseizinger reviewed (commented):
A few more comments but otherwise LGTM.

This is almost ready to merge, thank you! :)

John-LittleBearLabs committed:
not From

John-LittleBearLabs committed:
cargo fmt

thomaseizinger reviewed (approved):
LGTM!

I'll let @mxinden also have a look as he wrote the original issue!

John-LittleBearLabs committed:
clippy

John-LittleBearLabs committed:
changelog

thomaseizinger commented on misc/metrics/CHANGELOG.md:
```diff
@@ -1,5 +1,7 @@
 # 0.10.0 [unreleased]
 
+- Added `protocol_stack` metrics.
```
All entries apart from dependency bumps should also mention the PR they come from. See some of the other examples in this file :)

Plus, Max just pushed out 0.49, so this will need an update to the master branch and an according version bump (patch-level) please!

John-LittleBearLabs committed:
Merge remote-tracking branch 'origin/master' into addrstak_metric

John-LittleBearLabs committed:
Version bump.

thomaseizinger commented on misc/metrics/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+# 0.10.1
```
```suggestion
# 0.10.1 [unreleased]
```

thomaseizinger commented on misc/metrics/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+# 0.10.1
+
+- Added `protocol_stack` metrics. See [PR 2982].
```
Missing link for this reference.

thomaseizinger reviewed (approved):
Nice work!

@mxinden Thoughts from your end?

John-LittleBearLabs committed:
PR comment

mxinden commented on misc/metrics/examples/metrics/main.rs:
```diff
@@ -109,8 +113,22 @@ fn main() -> Result<(), Box<dyn Error>> {
 }
+
+impl Behaviour {
+    fn new(local_key: libp2p::identity::Keypair) -> Self {
```
How about passing the public key only here?

mxinden commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,86 @@
+    address_stack: String,
+}
+impl Label {
+    pub fn for_multi_address(ma: &Multiaddr) -> Self {
```
```suggestion
    pub fn from_multiaddress(ma: &Multiaddr) -> Self {
```

I don't think using `for_` is a convention. Am I missing something?

John-LittleBearLabs replied:
No, you're right that that is more idiomatic for rust and so it makes more sense. I'll be implementing From as mentioned below.

I think I landed on "for" without really thinking about it just because if I were describing the result/purpose of this method in English prose I would've likely used "for", but I'm not writing English prose here - I'm writing rust.

mxinden commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,86 @@
+pub struct Label {
+    address_stack: String,
+}
+impl Label {
```
Why not implement `From` for `Label`?

``` rust
impl From<&Multiaddr> for Label {
```

thomaseizinger replied:
I think we had this at some point and I recommended to move away from it. Unless we are actually using `From` as an abstraction somewhere, using a generic function has downsides in terms of clarity. It also encourages the use of `.into` at callsites which makes it hard to see, what this is actually being converted into.

John-LittleBearLabs replied:
Ah yes, that sounds familiar. I probably resolved & hid the comment when I committed the recommended change, but it's probably still above somewhere.

Personally I think I prefer the From approach in this case - since the struct has such an incredibly narrow usage I don't think there's too much confusion about what `into` is doing. But of course I'll go with whatever the consensus is.

thomaseizinger replied:
`From` would be nice if we could change the signature of `get_or_create` to accept an `impl Into<S>` so you don't have to call `.into`.

Unfortunately we can't do that without requiring ownership on _every_callsite which would require allocations even though we only need ownership inside the `get_or_create` function for the first call. Meaning a lot of these allocations would be wasted.

I'd be in favor of not obfuscating what we are converting here:

1. Using `.into` makes it hard to find all usages of `Labels`. Currently, I have to delete the `From` impl and see where the compiler fails.
2. I find this:
```rust
                    self.listen_addresses
                        .get_or_create(&protocol_stack::Labels::new(&listen_addr))
                        .inc();
```
much clearer to read than this:
```rust
                    self.listen_addresses
                        .get_or_create(&listen_addr.into())
                        .inc();
```
The former I can glance over whereas with the latter, my brain stops and is like: "Hang on, `into()` what?". And I have to recall the function signature of `get_or_create`, check the type of `listen_addr` and search for `From` implementations to understand what is happening here.

3. If we ever need more data for the labels, `From` would need to be using a tuple at which point we need a constructor anyway.
4. Out of principle, I think we should not be depending on abstractions (`From`), if we don't use them. Traits are only useful in generic code, so unless we can make a function that uses `T: Into` where we can pass a `&Multiaddr`, I think this doesn't qualify.

mxinden commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,86 @@
+use prometheus_client::encoding::text::Encode;
+
+#[derive(Encode, Hash, Clone, Eq, PartialEq)]
+pub struct Label {
```
```suggestion
pub struct Labels {
```
Nit pick, a `struct` is a set of labels.

mxinden commented on misc/metrics/src/swarm.rs:
```diff
@@ -156,8 +157,10 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
             }
-            libp2p_swarm::SwarmEvent::IncomingConnection { .. } => {
-                self.connections_incoming.inc();
+            libp2p_swarm::SwarmEvent::IncomingConnection { send_back_addr, .. } => {
```
How about doing the same for all other `SwarmEvent`s? E.g. `ConnectionEstablished`?

John-LittleBearLabs replied:
I added ConnectionEstablished, but do you really want to do _all_ of them, even things like `connected_to_banned_peer` ?

````
$for u in http://127.0.0.1:37857/metrics http://127.0.0.1:34293/metrics ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 6
libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
````

John-LittleBearLabs replied:
Added what I easily can, but these are triggered by events that don't seem to carry the an obvious multiaddr to key off of in the event struct itself:

dial_attempt
outgoing_connection_error
listener_error

These copy-pasted outputs are from the example running in my branch. Using exclusively tcp multiaddrs, hence the grep to find metrics keyed off protocol.
````
$for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
````

thomaseizinger replied:
> Added what I easily can, but these are triggered by events that don't seem to carry the an obvious multiaddr to key off of in the event struct itself:
> 
> - dial_attempt
> - outgoing_connection_error
> - listener_error

Yeah that is fine.

> These copy-pasted outputs are from the example running in my branch. Using exclusively tcp multiaddrs, hence the grep to find metrics keyed off protocol.
> 
> ```
> $for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done
> libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
> libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
> libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
> ```

We don't seem to be consistent in the naming. I'd actually prefer the "protocol"-inspired naming. How about we just call this label `protocols`:

```bash
$ for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done

libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{protocols="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",protocols="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",protocols="/ip4/tcp"} 1
```

John-LittleBearLabs replied:
I agree, @thomaseizinger . Now have:

```bash
$for u in http://127.0.0.1:{39507,42089}/metrics  ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{protocols="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",protocols="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",protocols="/ip4/tcp"} 1
```

mxinden commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,86 @@
+//  protocols when dependency version changes.
+trait MultiaddrExt {
+    fn protocol_stack(&self) -> String;
+}
```
I still need to cut a new release of `multiaddr`. Sorry for the delay.

mxinden reviewed (dismissed):
This is already proven useful, see https://github.com/libp2p/rust-libp2p/pull/2289#issuecomment-1299304244.

Couple of comments. How about expanding this beyond the two metrics you currently improve?

John-LittleBearLabs committed:
Merge remote-tracking branch 'origin/master' into addrstak_metric

John-LittleBearLabs committed:
A couple PR comments.

John-LittleBearLabs committed:
Use protocol stack for a few more metrics.

John-LittleBearLabs committed:
metric label naming consistency

John-LittleBearLabs committed:
remove impl From

thomaseizinger commented on misc/metrics/CHANGELOG.md:
```diff
@@ -16,6 +16,10 @@
 
 - Update to `libp2p-gossipsub` `v0.43.0`.
 
+- Added `protocol_stack` metrics. See [PR 2982].
```
```suggestion
- Add `protocol_stack` metrics. See [PR 2982].
```

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+pub struct Labels {
+    protocols: String,
+}
+impl Labels {
```
I am surprised the formatter doesn't put a newline here?

John-LittleBearLabs replied:
Double checked and the formatter doesn't. :shrug: 

Not sure these two blocks are distinct & separate enough to conceptually justify a blank line, but if that's how folks are used to reading it I'll do the consistent thing so it doesn't stand out (as it does not deserve to stand out).

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+pub trait MultiaddrExt {
+    fn protocol_stack(&self) -> String;
+}
+impl MultiaddrExt for Multiaddr {
```
Same here.

John-LittleBearLabs replied:
Added one.

thomaseizinger commented on misc/metrics/src/protocol_stack.rs:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+pub fn tag(proto: Protocol) -> &'static str {
```
```suggestion
fn tag(proto: Protocol) -> &'static str {
```

thomaseizinger reviewed (approved):
Lets merge it! Thanks for sticking to this and sorry for the back and forth in the review!

I think clippy still wants to have a word with you, otherwise, I am happy to merge this :)

John-LittleBearLabs committed:
PR comments & clippy

mxinden reviewed (commented):
Thanks for the follow-ups here @John-LittleBearLabs.

Once https://github.com/multiformats/rust-multiaddr/pull/62 is merged, and released and used in this pull request, this is good to go from my end.

John-LittleBearLabs committed:
Merge from master, update protocol_stack to use upstream. Should have been 2 different commits.

John-LittleBearLabs committed:
Fix a clippy and a no-default-feature use stmt

thomaseizinger committed:
Move `Labels` to usage site

The labels are specific to the metric they are used it whereas
the `protocol_stack` module is more generic and shouldn't have
a dependency on prometheus.

mergify[bot] committed:
Merge branch 'master' into addrstak_metric

mxinden reviewed (approved):
Thanks! Good to go from my side.

thomaseizinger committed:
Merge branch 'master' into addrstak_metric
